tuf/keydb.py

```diff
@@ -111,6 +116,44 @@ def create_keydb_from_root_metadata(root_metadata, repository_name='default'):
   else:
     create_keydb(repository_name)
 
+
+  if targets_map_file is not None:
+    tuf.formats.TARGETS_MAPFILE_SCHEMA.check_match(targets_map_file)
+
+    for junk, key_metadata in six.iteritems(targets_map_file['keys']):
+      if key_metadata['keytype'] in _SUPPORTED_KEY_TYPES:
+        # 'key_metadata' is stored in 'KEY_SCHEMA' format.  Call
+        # create_from_metadata_format() to get the key in 'RSAKEY_SCHEMA' format,
+        # which is the format expected by 'add_key()'.  Note: The 'keyids'
+        # returned by format_metadata_to_key() include keyids in addition to the
+        # default keyid listed in 'key_dict'.  The additional keyids are
+        # generated according to securesystemslib.settings.HASH_ALGORITHMS.
+
+        # The repo may have used hashing algorithms for the generated keyids that
+        # doesn't match the client's set of hash algorithms.  Make sure to only
+        # used the repo's selected hashing algorithms.
+        hash_algorithms = securesystemslib.settings.HASH_ALGORITHMS
+        securesystemslib.settings.HASH_ALGORITHMS = key_metadata['keyid_hash_algorithms']
+        key_dict, keyids = securesystemslib.keys.format_metadata_to_key(key_metadata)
+        securesystemslib.settings.HASH_ALGORITHMS = hash_algorithms
+
+        try:
+          for keyid in keyids:
+            # Make sure to update key_dict['keyid'] to use one of the other valid
+            # keyids, otherwise add_key() will have no reference to it.
+            key_dict['keyid'] = keyid
+            add_key(key_dict, keyid=None, repository_name=repository_name)
+
+        # Although keyid duplicates should *not* occur (unique dict keys), log a
+        # warning and continue.  However, 'key_dict' may have already been
+        # adding to the keydb elsewhere.
+        except tuf.exceptions.KeyAlreadyExistsError as e: # pragma: no cover
+          logger.warning(e)
+          continue
+
+      else:
+        logger.warning('Root Metadata file contains a key with an invalid keytype.')
+
```

Comment:
Please try not to copy-paste huge blocks like these (especially if they include dubious `keyid_hash_algorithm`-code that is prone to change .. see https://github.com/secure-systems-lab/securesystemslib/pull/227). :)

Why not just merge the two dictionaries `targets_map_file['keys']` and `root_metadata['keys']` into a new dict and then iterate over that one? We need to be careful though and define (and implement) the desired behavior for when  there are matching keyids in both dicts.

Reply:
I removed the repeated block in the latest commit. When merging the dictionaries, I chose to prioritize keys from root metadata if there is a conflict. With the current implementation's use of sha2-256 for all keyids, this conflict should not occur, but it is something that might require some more thought.